Mathlib/NumberTheory/JacobiSum/Basic.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2024 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import Mathlib.NumberTheory.MulChar.Basic
+import Mathlib.Algebra.Module.BigOperators
+
+/-!
+# Jacobi Sums
+
+This file defines the *Jacobi sum* of two multiplicative characters `χ` and `ψ` on a finite
+commutative ring `R` with values in another commutative ring `R'`:
+
+`jacobiSum χ ψ = ∑ x : R, χ x * ψ (1 - x)`
+
+(see `jacobiSum`) and provides some basic results and API lemmas on Jacobi sums.
+
+## References
+
+We essentially follow
+* [K. Ireland, M. Rosen, *A classical introduction to modern number theory*
+   (Section 8.3)][IrelandRosen1990]
+
+but generalize where appropriate.
+
+This is based on Lean code written as part of the bachelor's thesis of Alexander Spahl.
+-/
+
+open BigOperators Finset
+
+/-!
+### Jacobi sums: definition and first properties
+-/
+
+section Def
+
+namespace Finset
+
+-- Is this useful enough to be non-private?
+private lemma sum_eq_sum_one_sub {R M : Type*} [Ring R] [Fintype R] [AddCommMonoid M] (f : R → M) :
+    Finset.sum univ f = Finset.sum univ fun x ↦ f (1 - x) := by
+  refine Fintype.sum_bijective (1 - ·) (Function.Involutive.bijective ?_) _ _ fun x ↦ ?_
+  · simp only [Function.Involutive, sub_sub_cancel, implies_true]
+  · simp only [sub_sub_cancel]
+
+end Finset
+
+-- need `Fintype` instead of `Finite` for `Finset.sum` etc.
+variable {R R' : Type*} [CommRing R] [Fintype R] [CommRing R']
```

Comment:
I thought that the fashion now was `[Finite R]`?

Reply:
I think I tried that, and then `Finset.sum` complained...

Reply:
I think it is because `Finset.univ` requires a `Fintype` instance.

Reply:
`attribute [local instance] Fintype.ofFinite` fixes this (or `letI := Fintype.ofFinite F` in lots of places)

Reply:
My understanding is that when it is needed for a statement or definition, requiring `Fintype` should be OK, and if it is just needed in a proof, requiring `Finite` is preferred (and then using `Fintype.ofFinite` in the proof); by analogy with `Decidable`. If desired, one can always use `Fintype.ofFinite` in either way when using the results.

Note that using `Finite` instead of `Fintype`  makes `jacobiSum` noncomputable and does not allow proofs like
```lean
example : jacobiSum ZMod.χ₄ ZMod.χ₄ = 0 := rfl
```
or
```lean
example : jacobiSum ZMod.χ₄ ZMod.χ₄ = 0 := by decide
```

So I think keeping `Fintype` is better here.